datasource/src/main/scala/quasar/physical/mongo/decoder.scala

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.physical.mongo
+
+import slamdata.Predef._
+import qdata._
+import qdata.QType._
+import quasar.contrib.std.errorImpossible
+
+import java.time._
+import java.util.{Map, Iterator}
+
+import eu.timepit.refined.auto._
+import org.bson._
+import spire.math.Real
+
+object decoder {
+  @SuppressWarnings(Array("org.wartremover.warts.Var"))
+  class BsonDocumentCursor(document: BsonDocument) {
+
+    val iterator: Iterator[Map.Entry[String, BsonValue]] =
+      document.entrySet().iterator()
+
+    var entry: Option[Map.Entry[String, BsonValue]] = None
+
+    def hasNext(): Boolean = {
+      try {
+        entry = Some(iterator.next())
+        true
+      } catch {
+        case _: Throwable => false
+      }
+    }
+
+    def stepObject(): BsonDocumentCursor = this
+
+    def getObjectKeyAt(): String = entry.getOrElse(errorImpossible).getKey()
+
+    def getObjectValueAt(): BsonValue = entry.getOrElse(errorImpossible).getValue()
+
+  }
+
+  val qdataDecoder: QDataDecode[BsonValue] = new QDataDecode[BsonValue] {
+    override def tpe(bson: BsonValue): QType = bson.getBsonType() match {
+      case BsonType.DOCUMENT => QObject
+      case BsonType.ARRAY => QArray
+      case BsonType.STRING => QString
+      case BsonType.INT32 => QLong
+      case BsonType.INT64 => QLong
+      case BsonType.DOUBLE => QDouble
+      case BsonType.DECIMAL128 => QReal
+      case BsonType.BOOLEAN => QBoolean
+      case BsonType.OBJECT_ID => QString
+      case BsonType.DB_POINTER => QString
+      case BsonType.BINARY => QString
+      case BsonType.DATE_TIME => QOffsetDateTime
+      case BsonType.SYMBOL => QString
+      case BsonType.REGULAR_EXPRESSION => QString
+      case BsonType.JAVASCRIPT => QString
+      case BsonType.NULL => QNull
+      case BsonType.TIMESTAMP => QOffsetDateTime
+      case BsonType.UNDEFINED => QNull
+      case BsonType.JAVASCRIPT_WITH_SCOPE => QString
+      case BsonType.MAX_KEY => QString
+      case BsonType.MIN_KEY => QString
+      case BsonType.END_OF_DOCUMENT => QNull
```

Comment:
Can we use it? I mean the fact it's regexp

Reply:
Yeah, this could be a use for `meta`. We did this when encoding to `EJson` where the `meta` value would be an object like `{ "type": "mongodb.regex" }` or similar.